src/Tools/Pagination/Paginator.php

```diff
@@ -51,6 +75,190 @@ public function __construct(
+        }
+
+        $queryAST = $query->getAST();
+        [
+            'queryHasGroupByClause' => $queryHasGroupByClause,
+            'queryHasHavingClause' => $queryHasHavingClause,
+            'rootEntityHasSingleIdentifierFieldName' => $rootEntityHasSingleIdentifierFieldName,
+            'queryCouldProduceDuplicates' => $queryCouldProduceDuplicates,
+            'queryCouldHaveToManyJoins' => $queryCouldHaveToManyJoins,
+        ]         = self::autoDetectQueryFeatures($query->getEntityManager(), $queryAST);
+
+        $paginator = new self($query, $queryCouldProduceDuplicates);
+
+        $paginator->queryStyleAutoDetectionEnabled = true;
+        $paginator->queryCouldHaveToManyJoins      = $queryCouldHaveToManyJoins;
+        // The following is ensuring the conditions for when the CountWalker cannot be used.
+        $paginator->useCountQueryOutputWalker = $queryHasHavingClause !== false
+            || $rootEntityHasSingleIdentifierFieldName !== true
+            || ($queryCouldHaveToManyJoins && $queryHasGroupByClause !== false);
+
+        return $paginator;
+    }
+
+    /**
+     * @return array{
+     *  queryHasGroupByClause: bool|null,
+     *  queryHasHavingClause: bool|null,
+     *  rootEntityHasSingleIdentifierFieldName: bool|null,
+     *  queryCouldProduceDuplicates: bool,
+     *  queryCouldHaveToManyJoins: bool,
+     * }
+     */
+    private static function autoDetectQueryFeatures(EntityManagerInterface $entityManager, Node $queryAST): array
+    {
+        $queryFeatures = [
+            // Null means undetermined
+            'queryHasGroupByClause' => null,
+            'queryHasHavingClause' => null,
+            'rootEntityHasSingleIdentifierFieldName' => null,
+            'queryCouldProduceDuplicates' => true,
+            'queryCouldHaveToManyJoins' => true,
+        ];
+
+        if (! $queryAST instanceof Query\AST\SelectStatement) {
+            return $queryFeatures;
+        }
+
+        $queryFeatures['queryHasGroupByClause'] = $queryAST->groupByClause !== null;
+        $queryFeatures['queryHasHavingClause']  = $queryAST->havingClause !== null;
+
+        $from = $queryAST->fromClause->identificationVariableDeclarations;
+        if (count($from) > 1) {
+            return $queryFeatures;
+        }
+
+        $fromRoot = reset($from);
+        if (! $fromRoot instanceof Query\AST\IdentificationVariableDeclaration) {
+            return $queryFeatures;
+        }
+
+        if (! $fromRoot->rangeVariableDeclaration) {
+            return $queryFeatures;
+        }
+
+        $rootAlias = $fromRoot->rangeVariableDeclaration->aliasIdentificationVariable;
+        try {
+            $rootClassMetadata                                       = $entityManager->getClassMetadata($fromRoot->rangeVariableDeclaration->abstractSchemaName);
+            $queryFeatures['rootEntityHasSingleIdentifierFieldName'] = (bool) $rootClassMetadata->getSingleIdentifierFieldName();
+        } catch (MappingException) {
```

Comment:
Why is this necessary?

Reply:
You mean the catching of the `MappingException`? 

It's primarily needed because this is how `ClassMetadata::getSingleIdentifierFieldName()` communicates that "the identifier is not single column". One way to avoid the `try..catch` is to duplicate the if-conditions from `ClassMetadata::getSingleIdentifierFieldName()` in here, but having a code duplication here would introduce a maintenance cost, which I personally consider a bigger nuisance than having a `try..catch` block.

Also, `EntityManager::getClassMetadata()` throws that exception class when the `$className` is unrecognised. Instead of making an assumption that "there must be metadata for every class detected at this point of the code", I wanted to be more graceful and defensive, and chose to just catch the exception and assume that the "auto-detection" cannot be carried out.

Reply:
We usually avoid using exception for flow control, and I think that's because [they are costly](https://php.watch/articles/php-exception-performance).

I think I would add a dedicated method, so as to remove the need of casting the result to a boolean as well, even if it means a little duplication.

> Also, EntityManager::getClassMetadata() throws that exception class when the $className is unrecognised. Instead of making an assumption that "there must be metadata for every class detected at this point of the code", I wanted to be more graceful and defensive, and chose to just catch the exception and assume that the "auto-detection" cannot be carried out.

In what case will `$className` not be recognized and we wouldn't want to let the user know about that?